tensorflow-keras/src/main/java/org/tensorflow/keras/backend/K.java

```diff
@@ -0,0 +1,713 @@
+      x = tf.dtypes.cast(x, TFloat32.DTYPE);
+    }
+    if (axis == null) {
+      axis = allAxis(tf, x);
+    }
+    return tf.math.mean(x, axis, Mean.keepDims(keepDims));
+  }
+
+  // alias for mean
+  public static Operand reduceMean(Ops tf, Operand x, Operand axis, boolean keepDims) {
+    return mean(tf, x, axis, keepDims);
+  }
+
+  public static Operand maximum(Ops tf, Operand x, Operand y) {
+    y = tf.dtypes.cast(y, x.asOutput().dataType());
+    return tf.math.maximum(x, y);
+  }
+
+  public static <T extends TType> Operand<T> sqrt(Ops tf, Operand<T> x) {
+    DataType dType = x.asOutput().dataType();
+    Operand<T> zero = tf.dtypes.cast(tf.constant(0), dType);
+    Operand<T> inf = tf.dtypes.cast(tf.constant(Float.POSITIVE_INFINITY), dType);
+    x = tf.clipByValue(x, zero, inf);
+    return tf.math.sqrt(x);
+  }
+
+  public static Shape merge(Shape a, Shape b) {
+    assert a.numDimensions() == b.numDimensions()
+        : String.format("Shapes %s and %s are incompatible", a, b);
+    long[] array = new long[a.numDimensions()];
+    for (int i = 0; i < a.numDimensions(); i++) {
+      if (a.size(i) != Shape.UNKNOWN_SIZE) {
+        if (b.size(i) != Shape.UNKNOWN_SIZE) {
+          assert a.size(i) == b.size(i) : String.format("Shapes %s and %s are incompatible", a, b);
+        }
+        array[i] = a.size(i);
+      } else {
+        array[i] = b.size(i);
+      }
+    }
+    return Shape.of(array);
+  }
+
+  // this is from nn in Python, I could not find it in the Java frameworks.
+  public static Operand sigmoidCrossEntropyWithLogits(Ops tf, Operand labels, Operand logits) {
+    Shape lablesShape = labels.asOutput().shape();
+    Shape logitsShape = logits.asOutput().shape();
+    Shape newShape = merge(lablesShape, logitsShape);
+
+    Operand zeros = tf.dtypes.cast(tf.zerosLike(logits), logits.asOutput().dataType());
+    Operand cond = tf.math.greaterEqual(logits, zeros);
+
+    Operand relu_logits = tf.select(cond, logits, zeros);
+    Operand neg_abs_logits = tf.select(cond, tf.math.neg(logits), logits);
+    return tf.math.add(
+        tf.math.sub(relu_logits, tf.math.mul(logits, labels)),
+        tf.math.log1p(tf.math.exp(neg_abs_logits)));
+  }
+
+  // TODO need to walk back identity until it hits something else
+  // not sure how to get the input nodes for the Operand.
+  private static Operand backtrackIdentity(Operand output) {
+    // while(!output.op().type().equals("Identity"))
+    //    output = output.op().output(0);
+    return output;
+  }
+
+  public static Operand binary_crossentropy(
+      Ops tf, Operand target, Operand output, boolean fromLogits) {
+    if (fromLogits) {
+      return sigmoidCrossEntropyWithLogits(tf, target, output);
+    }
+
+    if (!(output instanceof Variable) && (!tf.scope().env().isEager())) {
+      // output = backtrackIdentity(output); // TODO - this does not work, goes infinite loop
+      if (output.op().type().equals("Sigmoid")) {
```

Comment:
I found "Sigmoid" in tensorflow-core/tensorflow-core-api/src/bazel/api_def/api_def_Sigmoid.pbtxt.
```
op {
  graph_op_name: "Sigmoid"
  endpoint {
    name: "math.Sigmoid"
  }
}
```
I am not sure how these files are parsed, but it seems the String name could be used to create a `public static final String` in the generated Java file.

Reply:
The `graph_op_name` is processed C++ code in the `tensorflow-core-api/src/bazel/op_generator` directory.
I modified `source_writer.cc`, `source_writer.h` to add a method `WriteFieldWithInitializer` and modified `op_generator.cc` to output the following static field for every generated Operator.

```
/** The name of this op, as known by TensorFlow core engine */
  public static final String OP_NAME = "SoftmaxCrossEntropyWithLogits";
```

This allows access to the graph operation name using, for example, `SoftmaxCrossEntropyWithLogits.OP_NAME` rather than using a raw string.
If everyone agrees with this approach,  I can push these changes

Reply:
I will be checking in the change for the bazel/op_generator C++ code.